File: adb/fastdeploy/deploypatchgenerator/patch_utils.cpp
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "patch_utils.h"

#include <stdio.h>

#include "adb_io.h"
#include "adb_utils.h"
#include "android-base/endian.h"
#include "sysdeps.h"

#include "apk_archive.h"

using namespace com::android;
using namespace com::android::fastdeploy;
using namespace android::base;

static constexpr char kSignature[] = "FASTDEPLOY";

APKMetaData PatchUtils::GetDeviceAPKMetaData(const APKDump& apk_dump) {
    APKMetaData apkMetaData;
    apkMetaData.set_absolute_path(apk_dump.absolute_path());

    std::string md5Hash;
    int64_t localFileHeaderOffset;
    int64_t dataSize;

    const auto& cd = apk_dump.cd();
    auto cur = cd.data();
    int64_t size = cd.size();
    while (auto consumed = ApkArchive::ParseCentralDirectoryRecord(
                   cur, size, &md5Hash, &localFileHeaderOffset, &dataSize)) {
        cur += consumed;
        size -= consumed;

        auto apkEntry = apkMetaData.add_entries();
        apkEntry->set_md5(md5Hash);
        apkEntry->set_dataoffset(localFileHeaderOffset);
        apkEntry->set_datasize(dataSize);
    }
    return apkMetaData;
}

APKMetaData PatchUtils::GetHostAPKMetaData(const char* apkPath) {
    ApkArchive archive(apkPath);
    auto dump = archive.ExtractMetadata();
    if (dump.cd().empty()) {
        fprintf(stderr, "adb: Could not extract Central Directory from %s\n", apkPath);
        error_exit("Aborting");
    }

    auto apkMetaData = GetDeviceAPKMetaData(dump);

    // Now let's set data sizes.
    for (auto& apkEntry : *apkMetaData.mutable_entries()) {
        auto dataSize =
                archive.CalculateLocalFileEntrySize(apkEntry.dataoffset(), apkEntry.datasize());
        if (dataSize == 0) {
            error_exit("Aborting");
        }
        apkEntry.set_datasize(dataSize);
    }

    return apkMetaData;
}

void PatchUtils::WriteSignature(borrowed_fd output) {
    WriteFdExactly(output, kSignature, sizeof(kSignature) - 1);
}

void PatchUtils::WriteLong(int64_t value, borrowed_fd output) {
    int64_t littleEndian = htole64(value);
    WriteFdExactly(output, &littleEndian, sizeof(littleEndian));
}

void PatchUtils::WriteString(const std::string& value, android::base::borrowed_fd output) {
    WriteLong(value.size(), output);
    WriteFdExactly(output, value);
}

void PatchUtils::Pipe(borrowed_fd input, borrowed_fd output, size_t amount) {
    constexpr static size_t BUFFER_SIZE = 128 * 1024;
    char buffer[BUFFER_SIZE];
    size_t transferAmount = 0;
    while (transferAmount != amount) {
        auto chunkAmount = std::min(amount - transferAmount, BUFFER_SIZE);
        auto readAmount = adb_read(input, buffer, chunkAmount);
        if (readAmount < 0) {
            fprintf(stderr, "adb: failed to read from input: %s\n", strerror(errno));
            error_exit("Aborting");
        }
        WriteFdExactly(output, buffer, readAmount);
        transferAmount += readAmount;
    }
}

File: adb/fastdeploy/deployagent/src/com/android/fastdeploy/PatchUtils.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.fastdeploy;

import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

class PatchUtils {
    public static final String SIGNATURE = "FASTDEPLOY";

    /**
     * Reads a 64-bit signed integer in Little Endian format from the specified {@link
     * DataInputStream}.
     *
     * @param in the stream to read from.
     */
    static long readLELong(InputStream in) throws IOException {
        byte[] buffer = new byte[Long.BYTES];
        readFully(in, buffer);
        ByteBuffer buf = ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN);
        return buf.getLong();
    }

    static String readString(InputStream in) throws IOException {
        int size = (int) readLELong(in);
        byte[] buffer = new byte[size];
        readFully(in, buffer);
        return new String(buffer);
    }

    static void readFully(final InputStream in, final byte[] destination, final int startAt,
            final int numBytes) throws IOException {
        int numRead = 0;
        while (numRead < numBytes) {
            int readNow = in.read(destination, startAt + numRead, numBytes - numRead);
            if (readNow == -1) {
                throw new IOException("truncated input stream");
            }
            numRead += readNow;
        }
    }

    static void readFully(final InputStream in, final byte[] destination) throws IOException {
        readFully(in, destination, 0, destination.length);
    }

    static void pipe(final InputStream in, final OutputStream out, final byte[] buffer,
            long copyLength) throws IOException {
        while (copyLength > 0) {
            int maxCopy = (int) Math.min(buffer.length, copyLength);
            readFully(in, buffer, 0, maxCopy);
            out.write(buffer, 0, maxCopy);
            copyLength -= maxCopy;
        }
    }
}
